Add "inline" option for unicode and console unit formats to give negative powers instead of fractions

olebole opened the pull request:
### Description

This PR adds inline versions of the unicode and console format representations for units. My primary interest is `unicode_inline`, but since I modified the console format as well, `console_inline` came out for (almost) free -- it is up to discussion whether it is useful enough to keep.

It looks like this:
```Python
>>> from astropy.constants import G
>>> print(f'{G:unicode_inline}')
6.6743e-11 m³ kg⁻¹ s⁻²
>>> print(f'{G:console_inline}')
6.6743e-11 m^3 kg^-1 s^-2
```
~~The change is done in the same way as in #2622 (which added `latex_inline`), to keep the same structure.~~

Im am also unsure what the use real case for the existing non-inline versions is as they are broken when used for quantities (IMO they should at least be disabled for formatting units, or switched to the proposed inline variants then):
```Python
>>> print(f'{G:unicode}')
6.6743e-11   m³  
 ─────
 kg s²
>>> print(f'{G:console}')
6.6743e-11   m^3  
 ------
 kg s^2
```

pllim commented:
Thanks! Is this related to #14386 ?

pllim commented:
Would need a change log and maybe a what's new too.

pllim commented:
Failures are definitely related, so those also need to be addressed.

olebole commented:
It is not related, and also not ready yet. It seems a bit more complicated than initially thought, because `Console._format_bases` needs the existing string (which was not the case in `Latex._format_bases`).

Maybe a better solution would be to add an `inline=False` class attribute instead, which set to **True** in the inline variants.

mhvk commented on astropy/units/format/console.py:
```diff
@@ -68,30 +69,51 @@ def to_string(cls, unit):
-                if len(negatives):
-                    if len(positives):
-                        positives = cls._format_unit_list(positives)
+                if cls.inline:
```
Wouldn't it make more sense to just overwrite the definition of `to_string()` in the definition of `ConsoleInline`?

Actually, possibly best is to make `ConsoleInline` the top class, and let `Console` inherit from it, since then it can use `super()` for the case that there is no `denominator`.

mhvk commented on astropy/units/format/latex.py:
```diff
@@ -82,7 +72,23 @@ def to_string(cls, unit):
 
             if len(unit.bases):
-                s += cls._format_bases(unit)
+                if cls.inline:
```
So, I guess I liked the old method somewhat better -- I think it is best either to have different classes that do their own thing, or to have a single class with an option.

mhvk reviewed (commented):
In principle, all for this! If implemented via subclasses, though, I think it is best to just override relevant methods rather than use an attribute.

That said, I'm actually not sure it is most sensible to add subclasses -- to me, these are very much stylistic changes for which keyword arguments or configuration items are more appropriate -- a bit like `np.printoptions`.

We've gone already a bit in that direction for `Quantity.to_string`, which has a `subfmt` option.  Should `Unit.to_string()` have something similar, which gets passed on to the format classes?

I worry in part because of seeing the test cases: why does 'generic' collect its negative units in parentheses and uses a solidus, while 'console' and 'unicode' use negative powers. That difference suggests there should be an option to choose, but do we really want an extra subclass for that too? Adding it as options to `format.to_string()` seems more logical...

mhvk commented:
@olebole - I made a separate issue about the formatting -- see #14398. My sense is that the easiest is for the format to default to inline -- which would likely be easier to implement if it is a keyword argument...

p.s. Sorry for making what looked like a simple PR complicated... But it would be nice to get it more or less right...

olebole commented:
I am answering in the main thread since it may make the discussion easier.

Overwriting `to_string()` was my first thought. However, then the structure would be different to `latex.py` which I didn't like. And a lot of code was duplicated in the overwritten class. To me, this looked more complicated than before.

Having an argument in `to_string()` would be the better way; however this would change the API, and I wanted to make a small change first. And going from there to `to_string()` options is trivial if everything is in one class , while with merging to classes it is more difficult.

If one would add an argument to `to_string()`:
```Python
    @classmethod
    def to_string(cls, unit, inline=False):
        […]
```
how would this be propagated to the user, f.e. in `print(f'{c:unicode]')`? And could this option set to True by default when formatting a quantity?

Another (more complicated) way could be to convert the formatters from classes to instances - this way one could have one unicode and one unicode_inline instance, and easily create others if needed (f.e. with central dots instead of spaces between the factors) -- however probably all use cases here could be better handled with subformats.

Can you point me to how quantities handle/specify subformats?

mhvk commented:
@olebole -- interesting! Just on the practical questions: arguments could be passed down from `Unit.to_string()` if it took arbitrary keyword arguments; see https://github.com/astropy/astropy/blob/f48987276a88fa785eb1964cc737aa371f989eed/astropy/units/core.py#L748-L759

I agree it would not help `Quantity.__format__`, though I feel that that one requires more thought anyway, of actually defining a useful extension to the standard formatting mini-language -- if we were going the route of keyword arguments, I'd just set the default to `inline=True`. But it would help with `Quantity.to_string()` -- as there again one could pass things down. Its implementation, including the currently rather kludgy `subfmt` is at https://github.com/astropy/astropy/blob/f48987276a88fa785eb1964cc737aa371f989eed/astropy/units/quantity.py#L1393-L1511

mhvk commented:
On the subclasses vs `cls.inline` -- in principle, subclassing should be fairly easy, since the non-inline version can use the inline version to typeset the nominator and denominator, and make those into a fraction. This is why I thought `Console` was more logically a subclass of `ConsoleInline` instead of vice versa. Anyway, I do still fear that this proliferation of subclasses is just going to hurt us, so let's think about the bigger picture first before considering implementation details...

olebole commented:
For `to_string()` I would go the way to add an `inline` argument; this would probably already resolve many use cases (like having the unit prettyprinted in a table header on console or on rtd/sphinxdoc output). Maybe we should limit the goal of this PR to this. For later, I like the idea of a mini language.
Then, the open questions for this PR would be:
 * If this is forwarded in `Unit.to_string()` like this:
    ```Python
    def to_string(self, format=unit_format.Generic, subfmt=None):
        […]
        return f.to_string(self, subfmt=subfmt) 
    ```
    even for formatters that don't implement subfmts. Or we generically forward `**kwargs`.
 * Shall we keep `latex_inline` or make it "`latex`, subfmt=`inline`"? The latter would be more consistent, but how would we add the display/inline subformat?
 * How much can we break compatibility? Making inline the default may be incompatible to current uses. Replacing `latex_inline` by  "`latex`, subfmt=`inline`" may break something. 
 * Or we just **define** for the moment that the `_` delimits units format and subformat, split them in `Unit.to_string()` and forward this:
    ```Python
    def to_string(self, format=unit_format.Generic):
        fmt = format.split('_', 1)
        f = unit_format.get_format(fmt[0])
        if len(fmt) == 1:       
            return f.to_string(self) 
        else:
            return f.to_string(self, subfmt=fmt[1]) 
    ```
    (and the query for the `Base.registry` should do the same)
    This would result in something that has the API identical to what we have, and it would be extendible. It would however promote the `_` as delimiter between unit format and subformat, which currently is used only by `latex_inline`.

mhvk commented:
I'll have to think a little more, in particular about the implications for `Quantity.to_string()`, but what you write sounds like a super plan!  I had independently wondered whether we could split on `_` to keep backward compatibility.

olebole commented:
I now again stepped back a bit and implemented just the basics; an "inline" option in `console`, `unicode`, `latex`, and used `inline=True` as default for `console` and `unicode`. Since `latex` multiline is used regularly (as formatting in IPython notebooks), I left the default there unchanged, and also left `latex_inline` for compatibility. The latter class has again trivial code (just overwriting the default), but this seems unavoidable for the moment.

I'd propose to take this as an acceptable intermediate step and to merge this. It only minimally changes the public API and postpones any "quantity format minilanguage", allowing to think here more about a more holistic approach. It also resolves somehoe #14398, because of the switch to inline formatting as default.
I am unsure how much we then would need the "multiline" options for console and unicode; maybe we wait here for user feedback, how much they are missed in the default formatting (I doubt that).

mhvk reviewed (commented):
Nice!  This works well, and I agree we might as well break it up into small steps. Only real comment is about a method I think is no longer used, plus a small suggestion.

mhvk commented:
pre-commit.ci autofix

olebole commented:
As the files were touched here anyway, I deleted/replaced the single `l`/`r` variable with something speaking,,,

mhvk reviewed (approved):
Thanks for changing those too -- definitely clearer! Since tests now all passed, I'll merge. Nice!
